integration-test/background/click-to-load-facebook.js

```diff
@@ -26,61 +70,74 @@ describe('Test Facebook Click To Load', () => {
         } catch (e) {}
     })
 
-    it('CTL: Should block Facebook requests by default', async () => {
+    it('CTL: Facebook request blocking/redirecting', async () => {
+        // Open the test page and start logging network requests.
         const page = await browser.newPage()
-
-        try {
-            await page.goto(testSite, { waitUntil: 'networkidle2', timeout: 10000 })
-        } catch (e) {
-            // timed out waiting for page to load, let's try running the test anyway
-            console.log(`Timed out: ${e}`)
+        const pageRequests = []
+        logPageRequests(page, pageRequests)
+
+        // Initially there should be a bunch of requests. The SDK should
+        // be redirected to our surrogate but otherwise Facebook requests should
+        // be blocked.
+        await page.goto(testSite, { waitUntil: 'networkidle0' })
+        await page.waitForNetworkIdle({ idleTime: 1000 })
+        {
+            const {
+                requestCount, blockCount, allowCount, facebookSDKRedirect
+            } = summariseFacebookRequests(pageRequests)
+
+            expect(facebookSDKRedirect.checked).toBeTrue()
+            expect(facebookSDKRedirect.alwaysRedirected).toBeTrue()
+            expect(requestCount).toBeGreaterThan(5)
+            expect(blockCount).toEqual(requestCount)
+            expect(allowCount).toEqual(0)
         }
 
-        // give it little time just to be sure (facebook widgets can take time to load)
-        await page.waitForTimeout(4000)
-        const fbRequestData = await page.evaluate(() => {
-            return {
-                requests: document.getElementById('facebook_call_count').innerHTML,
-                expectedCalls: document.getElementById('facebook_iFrames').innerHTML
-            }
+        // Once the user clicks to load the Facebook content, the SDK should be
+        // loaded and the content should be unblocked.
+        pageRequests.length = 0
+        const buttonCount = await page.evaluate(() => {
+            window.buttons =
+                Array.from(document.querySelectorAll('body > div'))
+                    .map(div => div.shadowRoot && div.shadowRoot.querySelector('button'))
+                    .filter(button => button)
+            return window.buttons.length
```

Comment:
hm, why are we returning count and not references to buttons? can we return array of "handles" and then perform clicks?

Reply:
so what would happen if you used evaluateHandle and returned window.buttons in this case? Wouldn't that be a handle to the buttons array?

Reply:
That would return a `JSHandle` of the array, instead of an `ElementHandle` of one of the buttons. I think the `JSHandle` can be passed to a function that you're evaluating in the page, but I don't think it can be used directly. For example, I just tried this:

```javascript
        for (let button of await page.evaluateHandle(() => window.buttons)) {
            await button.click()
        }
```

and that resulted in an exception `TypeError: (intermediate value) is not iterable`

Reply:
Gotcha, thanks for double-checking!